### PR TITLE
Add 2d grid-based movement demo

### PR DESCRIPTION
**Pull request by NathanLovato**
As promised, here's the grid-based movement demo!
Code-wise it's meant to show how to keep the Grid and the Pawns separate as much as possible.
Video overview of the code coming soon.

ezgif-4-3c04875463

### CODE REVIEW

File: 2d/grid_based_movement/grid/grid.gd [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+			return(node)
+
+
+func request_move(pawn, direction):
+	var cell_start = world_to_map(pawn.position)
+	var cell_target = cell_start + direction
+	
+	var cell_target_type = get_cellv(cell_target)
+	match cell_target_type:
+		-1:
```
Comment (bojidar-bg) [inline 2d/grid_based_movement/grid/grid.gd]: You might add `EMPTY = -1` to the end of `CELL_TYPES`.

File: 2d/grid_based_movement/pawns/actor.gd [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	if target_position:
+		move_to(target_position)
+		$Tween.start()
+	else:
+		bump()
+
+
+func get_input_direction():
+	return Vector2(
+		int(Input.is_action_pressed("ui_right")) - int(Input.is_action_pressed("ui_left")),
```
Comment (bojidar-bg) [inline 2d/grid_based_movement/pawns/actor.gd]: Maybe use `get_action_value` as per https://github.com/godotengine/godot/pull/16902?

File: 2d/grid_based_movement/pawns/actor.gd [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+func _process(delta):
+	var input_direction = get_input_direction()
+	if not input_direction:
+		return
+	update_look_direction(input_direction)
+
+	var target_position = Grid.request_move(self, input_direction)
+	if target_position:
+		move_to(target_position)
+		$Tween.start()
```
Comment (bojidar-bg) [inline 2d/grid_based_movement/pawns/actor.gd]: Shouldn't this be inside `move_to`?

File: 2d/grid_based_movement/pawns/actor.gd [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	$Pivot/Sprite.rotation = direction.angle()
+
+
+func move_to(target_position):
+	set_process(false)
+	$AnimationPlayer.play("walk")
+	var move_direction = (position - target_position).normalized()
+	$Tween.interpolate_property($Pivot, "position", move_direction * 32, Vector2(), $AnimationPlayer.current_animation_length, Tween.TRANS_LINEAR, Tween.EASE_IN)
+	$Pivot/Sprite.position = position - target_position
+	position = target_position
```
Comment (bojidar-bg) [inline 2d/grid_based_movement/pawns/actor.gd]: A few newlines and reorderings in this block of code might make it more readable.

File: 2d/grid_based_movement/pawns/actor.gd [hunk 5]
```diff
@@ -0,0 +1,51 @@
+extends 'pawn.gd'
```
Comment (bojidar-bg) [inline 2d/grid_based_movement/pawns/actor.gd]: Double-quotes.

**Comment by NathanLovato** [comment]
Thanks for the feedback! The move_direction vector was wrong as well in the pawn script, fixed it.

**Comment by NathanLovato** [comment]
Did one last touch of polish, added two comments, removed a useless line

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**